### ISSUE TITLE
MPRIS seems broken in KDE 5.16.5

### ISSUE THREAD

**Issue by mohemohe** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#3618]
**Checklist:**

* [x] I have checked that there are no issues with similar or the same content
  (I think similar in #3583 , but that issue in Gnome and not discussed for MPRIS specification.)

  **YOU SHOULD CHECK CLOSED ISSUES ASWELL**
* [x] I have checked the FAQ (https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/wiki/FAQ) and the answer I am looking for is not there
* [x] I have double checked and can reproduce the issue

**OS:**

Arch Linux (last updated in about 3:00 UTC)

**GPMDP Version:**

both:

- 4.6.1 from AUR
- 5332c05

**Issue Descriptions:**

After updating to KDE5.16.5, the media control applet no longer detects GPMDP.
According to KDE [changelog](https://kde.org/announcements/plasma-5.16.4-5.16.5-changelog.php), that I think requests more strict MPRIS specification.

> Players not compliant to MPRIS specification will no longer work with Media Controller applet. [Commit](https://cgit.kde.org/plasma-workspace.git/commit/?id=8eec2cfd36c4ecf9e2b13f4226ed49495271a1e2). Phabricator Code review D22596

**Steps to Reproduce:**

1. Play any music.
2. Media Controller applet has not reaction.
3. Sometimes not working media control keys without MPRIS.

**Comment by ngseer** [comment]
The issue is still presented in Arch with Plasma 5.17

Here are some related logs:
```
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "Supported MimeTypes" as D-Bus type "<unknown>" but it should be D-Bus type "as"
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "Supported UriSchemes" as D-Bus type "<unknown>" but it should be D-Bus type "as"
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "MaximumRate" as D-Bus type "y" but it should be D-Bus type "d"
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "MinimumRate" as D-Bus type "y" but it should be D-Bus type "d"
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "Position" as D-Bus type "y" but it should be D-Bus type "x"
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "Rate" as D-Bus type "y" but it should be D-Bus type "d"
"org.mpris.MediaPlayer2.google-play-music-desktop-player" exports "Volume" as D-Bus type "y" but it should be D-Bus type "d"
```

According to the commit mentioned before, `SupportedUriSchemes` and `SupportedUriSchemes` values are mandatory.
I think it shouldn't take long to fix. The question is, which values should be used. My guess is `["audio/mpeg"]` for `MimeTypes` since GPM [seems to use MP3](https://support.google.com/googleplaymusic/answer/1100462) and `["http", "https"]` for `UriScheme` since there are no any other schemes in requests GPM makes. I'm not 100% sure about these but it may be enough to fix the issue.
I'm going to try it myself but don't have much time atm so if anyone else wants to help it would be appreciated.

**Comment by jostrander** [comment]
Yeah @ngseer that seems to be the solution. We've always left those out because there's really no way to be "compliant" being that this player doesn't play any user generated media from the local system. Will PR it in.

**Comment by jostrander** [comment]
Can anyone confirm that https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/commit/a1d6d348b8fa6a0f5db3b473aa60384581e77d6d fixes the issue?

**Comment by TurboGit** [comment]
@jostrander : I tried the solution hinted by @ngseer (maybe wrongly) and it did not solve the issue on my side.

**Comment by sollidsnake** [comment]
It works for me if I run `killall plasmashell && plasmashell` after starting plasma

**Comment by tristanlins** [comment]
It works for me, when I cherry pick the fix into the master branch. (Without killing plasma)
At least what I can see at first glance.

![Screenshot_20191104_090738](https://user-images.githubusercontent.com/343404/68106782-fa7beb80-fee2-11e9-9be0-7f5784dd5184.png)

<details>

Distributor ID: neon
Description:    KDE neon User Edition 5.17
Release:        18.04
Codename:       bionic

Linux thinkpad 5.3.0-19-generic

plasmashell 5.17.2
Qt: 5.13.1
KDE Frameworks: 5.63.0
kf5-config: 1.0

</details>

**Comment by ngseer** [comment]
@TurboGit could you please provide an output of `journalctl --no-pager -b -g org.mpris.MediaPlayer2.google-play-music-desktop-player` after applying the solution?

**Comment by ngseer** [comment]
@TurboGit it seems that you didn't run GPMDP on this boot. Could you please run it, play some media and then get an output of that command again? 
Sorry for not being clear in the first place.

**Comment by TurboGit** [comment]
@ngseer : No, it was clear at least that's what I understood to be sure i just restarted and launched the player and while a music was played I again run the command:

```
$ journalctl --no-pager -b -g org.mpris.MediaPlayer2.google-play-music-desktop-player
-- Logs begin at Thu 2019-11-07 18:34:37 CET, end at Thu 2019-11-07 18:39:58 CET. --
-- No entries --
```

Now I grepped _mpris_ on /var/log/messages and found some occurrences with Player instead of google-play-music-desktop-player, and using this I get:

```
$ journalctl --no-pager -b -g org.mpris.MediaPlayer2.Player
-- Logs begin at Thu 2019-11-07 18:34:37 CET, end at Thu 2019-11-07 18:41:23 CET. --
Nov 07 18:37:10 pascal gnome-shell[2325]: Ignored exception from dbus method: Gio.IOErrorEnum: La méthode « org.mpris.MediaPlayer2.Player.PlayPause » a renvoyé le type « (u) », mais « () » était attendu
Nov 07 18:37:12 pascal gnome-shell[2325]: Ignored exception from dbus method: Gio.IOErrorEnum: La méthode « org.mpris.MediaPlayer2.Player.PlayPause » a renvoyé le type « (u) », mais « () » était attendu
```

Does that helps ?

**Comment by TurboGit** [comment]
Just in case it is needed, here I have:
- Debian sid
- GNOME Shell 3.34

**Comment by ngseer** [comment]
@TurboGit yes, thank you. It is important to make sure it'd be fixed for all common setups.
I'm going to check a couple of things on my side and will come back soon with some summary.

If you'd have some spare time, I would also appreciate some logs for builds from both `master` and your fork as well.

**Comment by TurboGit** [comment]
@ngseer : sure, I'll do my best to help fixing that properly.

**Comment by ngseer** [comment]
Okay, here’s what I’ve got after some experiments with different builds on **Plasma 5.17**.

**tl;dr** I was able to make player controls working with @jostrander fix but it still may require some further investigations (see the conclusion of this post).

First of all, it seems that any build that doesn’t respect MPRIS specs won’t work, as it was mentioned by Plasma devs.
As for a1d6d34, indeed, as @sollidsnake mentioned, I had to restart plasmashell every time I reboot the system to make the controls work which, of course, didn’t look very nice thing to do.
So I’ve been trying different options and finally got some results.

**Good news**: it seems that restarting DBus service (followed by a reboot, maybe) fixes the issue completely. I was able to reproduce it on my second computer.
**Bad news**: understanding the reason of this peculiar behaviour requires more complex DBus knowledge than I have. My guess is, it stores imported values in some kind of cache. 
I've also tried to run the build from `master` afterwards and got everything turned to its original state: media controls doesn’t work with the fixed version until you restart plasmashell.

Considering all said, here’s my **conclusion**:

- a1d6d34 by @jostrander fixes the MPRIS specs incompatibility. Syslog errors are gone and media controls are working after some manipulations with the system that have nothing to do with the GPMDP itself.
- I’d like someone else to test and confirm that DBus restart (`systemctl restart dbus.service` as root for systemd-based distributions) followed by a reboot resolves the issue completely. **Be careful** since a lot of apps will not expect DBus restarted so make sure to save all your progress before these experiments.
- Another good thing to check is if the fresh installs on systems that never had GPMDP installed require these DBus manipulations at all.

I also believe that a1d6d34 may be merged right now since obeying the specs is a right thing to do regardless of the issue.

**Comment by TurboGit** [comment]
Ok, I did test all that:
- Get current master
- Apply fix in commit a1d6d34
- Clean everything locally (rm -fr build node_modules dist)
- Did a systemctl restart dbus.service (which has crashed by GNOME session :)
- Rebooted then
- Relaunched the build

And no, there is no MPRIS on GNOME.

I then did apply my patch again and rebuilt again.

And yes, I now do have MPRIS on GNOME.

So the commit a1d6d34 is not an issue on GNOME but it does not fix the MPRIS issue.

Let me know if I can help more. Thanks for all the investigations.

**Comment by jostrander** [comment]
Just tested a1d6d34 on Arch with gnome 3.34.1 and mpris appeared to work. Did not appear to work on master however. Did an `rm -rf node_modules` and `yarn && yarn run build` between branches. I'm not sure why this wouldn't be working for you @TurboGit.

**Comment by TurboGit** [comment]
Not sure why, I tried again:

```
$ rm -fr node_modules build
$ git checkout a1d6d348b8fa6
$ npm install ; npm run build ; npm start
```

No MPRIS on the center notification of GNOME Shell. I'm using version 3.34.1+git20191024-1 at the moment.

**Comment by TurboGit** [comment]
With the same procedure when I apply my "fix" it works. Can't tell more as I said previously I'm no expert on Node.js.

**Comment by jostrander** [comment]
@TurboGit I think I  might know why. `npm` doesn't have a package-lock in this repo, you have to use yarn to get the right dependencies from the `yarn.lock` file. Try using `yarn` or some package-lock to yarn.lock converter to get the right dep set.

**Comment by TurboGit** [comment]
Same with yarn:

```
$ rm -fr node_modules build
$ yarn && yarn run build && yarn start
```
MPRIS ok with my fix, not ok when building from commit a1d6d348b8fa6.

**Comment by ngseer** [comment]
@TurboGit just to be sure: as your referring #3661 as “my fix“?

The main difference here is the `mpris-service` library fork. It [differs a lot from the origin](https://github.com/GPMDP/mpris-service/compare/master...dbusjs:master) now. 
And the most noticeable change at the origin is a migration from the `node-dbus` to the `node-dbus-next`. That might be a good place to look for the cause of this problem.
E.g. one may try to build GPMDP against some [older versions](https://www.npmjs.com/package/mpris-service?activeTab=versions) of `mpris-service` to pinpoint the breaking changes.

In any case, returning to usage of a strongly outdated library looks like a bad idea to me.
I agree with @jostrander that we should make it work without resorting to using some old code that wasn't maintained for a long time. 

Maybe it deserves a separate issue after all since it's not related to Plasma anymore.

**Comment by TurboGit** [comment]
@ngseer : yes, I'm referring to #3661 as "my fix".

**Comment by ngseer** [comment]
@jostrander any hope we'll get #3673 merged soon?

**Comment by jostrander** [comment]
#3673 was merged, and 4.7.1 was released. Can anyone confirm this was fixed?

**Comment by TurboGit** [comment]
@jostrander : No I had already tested that and it does not fix the issue on my system. I'm still applying the fix I had proposed which works for me. I have no better idea !

**Comment by ngseer** [comment]
> #3673 was merged, and 4.7.1 was released. Can anyone confirm this was fixed?

**Arch with Plasma 5.17**, two different setups:
Media controls load only after Plasma restart (as originally mentioned by @sollidsnake).
Restarting DBus service doesn't help anymore.

**Kubuntu 19.10 Live**:
Fresh install works just fine. 
But if I install 4.7.0 first, run it and then upgrade to 4.7.1, it behaves just like in the Arch case (media controls don't load until I restart Plasma).

System log errors are gone in 4.7.1 and quick check with DBus viewer indicates that all metadata is exported just as expected.
From that perspective it seems that GPMDP does its part just fine and there's some bug/unexpected behavior in Plasma that still causes these problems with media controls.

I'll let you know if I'll find a solution.

**Comment by tristanlins** [comment]
On **KDE neon User Edition 5.17** it differs depend on the installation.
When using the snap package, it does not work.
When using the dep package, it works.

(plasma restart is not required anyway)

**Comment by aichingm** [comment]
> #3673 was merged, and 4.7.1 was released. Can anyone confirm this was fixed?

Also **Arch with plasma 5.17** GPMDP **4.7.1** works partly once I restarted the system. 

The only thing which does not work is the progress slider it ticks along fine but changing it does not seam to work. It changes the time values but the played track does not change position at all.

**Comment by tristanlins** [comment]
I switched back from the deb package to the snap package because the deb version verifiably causes my desktop to crash (KDE resets and I end up with the login mask again).
But there is no "indicator" of what exactly causes this.

I tried to watch the dbus, unfortunately I couldn't find anything useful. There is also nothing relevant to be found in the xsession-errors.

**Comment by ZOXSOCKS** [comment]
I'm also experiencing the above issue in Fedora 31 KDE Plasma 5.17.5.

I am monitoring this thread for updates! Thanks.

**Comment by ngseer** [comment]
Honestly, I'm totally confused about it now.
DBus imports seem just fine, I can control GPMDP through QDBus and third-party MPRIS controllers that use some different MPRIS implementation (e.g. `playerctl`).
It's plasma controller that still behaves strangely, sometimes it works without killing plasmashell, sometimes it doesn't.

I've been considering reporting a bug to the upstream but still don't have enough time to get any debug information. Maybe someone else can help with this.

**Comment by bhechinger** [comment]
I ended up on this issue after installing this (4.7.1) on Ubuntu Studio 21.04. The controls don't work and it doesn't do discord integration (even though I've enabled that).
